WIP: Atomic dependency handling for openqa-clone-job

Martchus opened the pull request:
* Very WIP, e.g. the actual handling of dependency parameters is still missing.
* See https://progress.opensuse.org/issues/103971#note-12 for how the API would be used.

Martchus commented:
I must say this was *way* easier than expected. I could mainly add new code without having to restructure any existing code except the `create` function itself (although most of its code has just been moved to the new function `_create_job` without much changes in the logic). Since the API remains backwards compatible all existing API tests pass unchanged.

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -219,60 +220,174 @@ sub overview {
+containing the removed parameters (where the keys are job suffixes).
+
+So "{foo => 1, bar:jobA => 2, baz:jobB => 3}" will be turned into "{foo => 1}"
+returnig "{jobA => {bar => 2}, jobB => {baz => 3}}".
```
s/returnig/returning/, better yielding

Martchus replied:
I'll change it to returning but yield is mainly used in the context of generators/co-routines so I wouldn't use it here.

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -219,60 +220,174 @@ sub overview {
-    my $params = $self->req->params->to_hash;
-    my $is_clone_job = delete $params->{is_clone_job} // 0;
+sub _eval_param_grouping ($params) {
+    my %grouped_params;
+    for my $param_key (keys %$params) {
+        my $job_suffix_start = rindex($param_key, ':');
+        next unless $job_suffix_start >= 0;
+        my $job_suffix = substr $param_key, $job_suffix_start + 1;
+        my $setting_key = substr $param_key, 0, $job_suffix_start;
+        $grouped_params{$job_suffix}->{$setting_key} = delete $params->{$param_key};
+    }
+    return \%grouped_params;
```
Something like this maybe

```suggestion
    map { ($job, $setting) = split /:/; $setting => {$job => $x{$_}} } grep { /:/ } keys %$params;
```

For a test:

```
perl -e '%x = ("foo" => 1, "bar:jobA" => 2, "baz:jobB" => 3); %h = map { ($job, $setting) = split /:/; $setting => {$job => $x{$_}} } grep { /:/ } keys %x; use Data::Dumper; print(Dumper(\%h))'
```

shows

```
$VAR1 = {
          'jobB' => {
                      'baz' => 3
                    },
          'jobA' => {
                      'bar' => 2
                    }
        };
```

Martchus replied:
* This breaks the deletion.
* This breaks if multiple `:` are present. The splitting should really only occur after the last colon.

But I might be able to make it use `map`.

Martchus replied:
This also breaks if you add `, "bar2:jobA" => 4` (which should result in multiple keys within `jobA`'s nested hash but doesn't.

okurz replied:
I am sure there is a way to use map to create the hash with correct values even if the same key appears multiple times but if we can't find it then leave it

Martchus replied:
But we're not creating "the hash" here. We're creating one hash containing an arbitrary number of nested hashes - and the nested hashes are not populated at once.

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -219,60 +220,174 @@ sub overview {
+    return undef unless $child_job_suffix;
+    my $job_suffixes = delete $job_settings->{$setting_key};
+    return undef unless defined $job_suffixes;
+    my $deps = $self->{_dependencies};
+    for my $parent_job_suffix (split /\s*,\s*/, $job_suffixes) {
+        push @$deps,
+          {
+            child_job_id => $child_job_suffix,
+            parent_job_id => $parent_job_suffix,
+            dependency => $type,
+          };
+    }
```
```suggestion
    my $deps = ($self->{_dependencies}, map { {
            child_job_id => $child_job_suffix,
            parent_job_id => $_,
            dependency => $type } } split /\s*,\s*/, $job_suffixes;
```

Martchus replied:
This makes no sense. `$self->{_dependencies}` is supposed to be populated here.

okurz replied:
oh, I see. I misread `my $deps = $self->{_dependencies} ` as deep copy which it isn't. So maybe

```suggestion
$self->{_dependencies} = ($self->{_dependencies}, map { {
            child_job_id => $child_job_suffix,
            parent_job_id => $_,
            dependency => $type } } split /\s*,\s*/, $job_suffixes);
```

because you can merge a second hash (here created by "map") into the first like `%h = (%h, %j)` or here more like `%h = (%h, map { … } @j)`

Martchus replied:
That's harder to read and unlikely to be more efficient. And it gets even uglier because you needed to make a new hash-ref (by wrapping in `[]` instead of `()`) and you needed to de-ref `$self->{_dependencies}` to retain the flat structure.

okurz replied:
I didn't invent it. I thought this is the way how to merge hashes in perl

Martchus replied:
I think of it more as "concatenation". "Merge" means to me something becomes part of something else but here we make something new and throw away the original parts.

okurz replied:
Yes, I meant that `%h = (%h, %j)` is the way to concatenate hashes in perl

Martchus commented on lib/OpenQA/JobDependencies/Constants.pm:
```diff
@@ -14,6 +16,8 @@ use constant DIRECTLY_CHAINED => 3;
 use constant DEPENDENCIES => (CHAINED, PARALLEL, DIRECTLY_CHAINED);
 use constant CHAINED_DEPENDENCIES => (CHAINED, DIRECTLY_CHAINED);
 
+our @EXPORT = qw(CHAINED PARALLEL DIRECTLY_CHAINED);
```
I'll check whether it works without it. The only goal here is to avoid `OpenQA::JobDependencies::Constants::…` all the time.

Martchus replied:
Without it I get

```
"CHAINED" is not exported by the OpenQA::JobDependencies::Constants module
"PARALLEL" is not exported by the OpenQA::JobDependencies::Constants module
"DIRECTLY_CHAINED" is not exported by the OpenQA::JobDependencies::Constants module
Can't continue after import errors at /hdd/openqa-devel/repos/openQA/lib/OpenQA/WebAPI/Controller/API/V1/Job.pm line 10.
```

Do you want me to keep using the `OpenQA::JobDependencies::Constants::…` everywhere?

okurz replied:
You would need to use

`use OpenQA::JobDependencies::Constants qw(CHAINED PARALLEL DIRECTLY_CHAINED);`

on the imports. I am ok if you use EXPORT but then better export all the symbols that you think are useful, not just the ones you just happen to need in this PR

okurz replied:
ok, fine

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -219,60 +220,174 @@ sub overview {
+=cut
 
+sub _eval_dependency ($self, $child_job_suffix, $job_settings, $setting_key, $type) {
+    return undef unless $child_job_suffix;
+    my $job_suffixes = delete $job_settings->{$setting_key};
+    return undef unless defined $job_suffixes;
+    my $deps = $self->{_dependencies};
+    for my $parent_job_suffix (split /\s*,\s*/, $job_suffixes) {
+        push @$deps,
+          {
+            child_job_id => $child_job_suffix,
+            parent_job_id => $parent_job_suffix,
+            dependency => $type,
+          };
+    }
+}
+
+=over 4
+
+=item _create_job()
+
+Creates a single job. The parameters $global_params and $job_specific_params will
+be merged ($job_specific_params takes precedence).
+
+=back
+
+=cut
+
+sub _create_job ($self, $global_params, $job_suffix = undef, $job_specific_params = {}) {
     # job_create expects upper case keys
-    my %up_params = map { uc $_ => $params->{$_} } keys %$params;
+    my %up_params = map { uc $_ => $global_params->{$_} } keys %$global_params;
+    $up_params{uc $_} = $job_specific_params->{$_} for keys %$job_specific_params;
+
     # restore URL encoded /
     my %params = map { $_ => $up_params{$_} =~ s@%2F@/@gr } keys %up_params;
-    if (!$params{TEST}) {
-        return $self->render(json => {error => 'TEST field mandatory'}, status => 400);
-    }
+    die "TEST field mandatory\n" unless $params{TEST};
```
Maybe this just looks weird in the diff but just to double check: Are you changing the error rendering back into a die? Didn't you just recently have a PR doing the opposite?

Martchus replied:
First, my other PR was about a whole different API. Second, this is still rendered as user-facing error because the controller code simply renders all `die` messages as such. (That isn't changed by my PR. A more fain-grained error handling would be nice but I consider it out of scope for now.)

okurz commented on lib/OpenQA/WebAPI/Controller/API/V1/Job.pm:
```diff
@@ -219,60 +220,174 @@ sub overview {
-                OpenQA::Scheduler::Client->singleton->wakeup;
+                if (keys %$grouped_params) {
+                    # create as many jobs as unique ":"-suffixes exist
+                    $self->_create_job($global_params, $_, $grouped_params->{$_}) for keys %$grouped_params;
+                }
+                else {
+                    # create a single job if there are no job-specific parameters
+                    $self->_create_job($global_params);
```
That additional level of indendation is a smell for me. How about making the additional parameters mandatory in `_create_job` and adding the single job parameter here as default in case `$grouped_params` is empty. Then you might be able to simplify the condition in https://github.com/os-autoinst/openQA/pull/4535/files#diff-6da53efb4105c5db577df2d5cd890fac0c8f83cd05a4b6a80bf5b067a4597fb2R307 as well

Martchus replied:
Not sure whether I follow. Should the for-loop go into `_create_job` (which should then likely be named `_create_jobs_internally`)? I also don't see whether this helps simplifying the condition. The condition is about whether `ids` or `id` is populated and the latter is still required for keeping backwards compatibility. Only if there is at least one suffix (even if there's still just a single job¹) the `ids` key should be populated.

¹Otherwise it is more inconsistent in my opinion, e.g. I'd needed to distinguish between `id` and `ids` in the clone script.

okurz replied:
> Not sure whether I follow. Should the for-loop go into `_create_job` (which should then likely be named `_create_jobs_internally`)?

I think so, yes. I haven't tried this myself yet to be sure this would be better, sorry.

Martchus replied:
Ok, I'll try it tomorrow and see whether it is better or not.

Martchus replied:
Moved into a separate function.

Martchus commented:
By the way, for better backwards compatibility I *could* make the splitting at `:` optional - so it would only happen if an extra parameter is specified (e.g. `multiple_jobs=1`). I refrained from that extra complexity because I think we don't use `:` in keys so existing usage shouldn't be impacted. (Note that this change actually still allows having `:` in keys as the splitting only happens after the last `:`. It just requires using the new syntax, e.g. by just adding another trailing `:`. That means one can still clone jobs with `:` setting keys, e.g. `TEST:FOO=bar` would be passed as `TEST:FOO:42=bar` by the clone script.)

Martchus commented:
I've rebased and added some documentation which hopefully makes the review a bit easier. I've also noticed that `openqa-cli` currently doesn't allow to pass parameters with `:` in it so I extended the regex (and corresponding tests).

kalikiana reviewed (approved):
I love the docs.

Martchus commented:
Thanks for the nice comment :-)

okurz reviewed (commented):
I haven't seen the doc changes at time of approval. I recommend to also extend the openqa-cli help correspondingly a bit and point to the users guide for details
